### ISSUE TITLE
vmMemoryOverheadPercent is causing over provision. It should be considered only after launching machines

### ISSUE THREAD

**Issue by bnetzi** [aws/karpenter-provider-aws#6611]
### Description

**Observed Behavior**:

vmMemoryOverheadPercent is causing karpenter to over provision nodes in case the memory requests are calculated accurately according to instance type.

For example, I'm using the i4i.32xlarge instance type which has 128vCPUS and 1024GB memory

Launched 4 pods, each with these configs:

    resources:
      limits:
        ephemeral-storage: 90Gi
        memory: 241620Mi
      requests:
        cpu: "31"
        ephemeral-storage: 5Gi
        memory: 240954Mi

Karpenter Should Launch only one as it fits in one node
But Karpenter actually launched two, as according to karpenter this node has 961470Mi, although in reality it has 1007423Mi Memory allocatable in k8s. (vmMemoryOverheadPercent default is  0.075 which sound low, but it is not really a percent, it's a factor so it is **7.25%!**)

Also - karpenter will not consolidate the nodes as he thinks there is no place for all the pods in one node, but when I kill the extra node, all the pods are now scheduled to same node without any issue.

This issue happened for us in large scale and caused us mass over provision as we are using large pods and we are trying to max the node resource allocation   

**Expected Behavior**:

Karpenter should launch according to the memory specs before the vmMemoryOverheadPercent and if the scheduler in use is failing to allocate (meaning pods are still pending)  - than it should launch more instances

**Reproduction Steps** (Please include YAML):
Create 4 pods with theses specs:

resources:
      limits:
        ephemeral-storage: 90Gi
        memory: 240954Mi
      requests:
        cpu: "31"
        ephemeral-storage: 5Gi
        memory: 240954Mi

And use only 128vCPU instances in the nodePool / provisioner

**Versions**:

Tested on version 0.32.6
- Chart Version:  0.32.6
- Kubernetes Version (`kubectl version`): 1.29

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by njtran** [comment]
Unfortunately, vmMemoryOverheadPercent is just an approximation, as Karpenter has no way to know exactly what the memory is without knowing the details of the AMI. The best advice we have is to test out Karpenter in lower environments with pods simulate what you're expecting to run in production, and ensure that Karpenter runs as expected. 

One way to solve this would be through [NodeOverlays](https://github.com/kubernetes-sigs/karpenter/pull/1305) if you're interested in that.

I see you also cut a docs issue. Feel free to cut a docs PR if you think this warrants better documentation. Since this isn't a problem we can solve, and there's an active RFC to address this, I'll go ahead and close this. Feel free to re-open if you want.

**Comment by bnetzi** [comment]
I'm opening it as I think this is a huge design flaw. The fact that karpenter can't know upfront does not mean it should assume a constant overhead for all instances types and for all node pools.
It should be at least be a factor that depends on instance size - 7% overhead of the memory is acceptable for small instances, but for large instances it creates a huge overhead. 
In my opinion the bare minimum would be to allow this to be configured per node pool. Node overlay seems like an ok solution too but it would be harder to configure, the node pool seems like a better place (although I do get why it is harder to implement it there.

Also - I still don't understand, why does karpenter needs to know upfront? it can launch an instance, check if it fits, and if not - kill it and launch another instance. Another option is to cache the allocatable memory for each instnaceType from each AMI used.

**Comment by bnetzi** [comment]
Hi @njtran do you want me to open a new issue or you can re-open this one? I don't have permission to re-open it myself
